### ISSUE TITLE
Story: Implement a Debug server for counterexample visualisation

### ISSUE THREAD

**Issue by keyboardDrummer** [boogie-org/boogie#539]
A Boogie project `DebugServer` is added that contains an API like:
```
public Task StartServer(Counterexample counterexample, PipeReader input, PipeWriter output);
```

That starts a server which implements the debug adapter protocol: https://microsoft.github.io/debug-adapter-protocol/

The server will support these features:
```
interface Capabilities {
  supportsConfigurationDoneRequest? = true;
  supportsRestartRequest? = true
  supportsStepBack?: boolean = true
  supportsGotoTargetsRequest?: boolean = true;
  supportsBreakpointLocationsRequest? = true
  supportsExceptionInfoRequest? = true
}
```

Debug steps need to be configured by Boogie clients, using assume commands of the form:
```
assume {:captureState} true
```

### Out of scope
Here is the list of features the debug server won't support:
```

interface Capabilities {
  supportsStepInTargetsRequest?: boolean = false;
  supportsSteppingGranularity? = false

  supportsFunctionBreakpoints? = false;
  supportsConditionalBreakpoints? = false;
  supportsHitConditionalBreakpoints? = false;
  supportsEvaluateForHovers? = false;
  exceptionBreakpointFilters?: ExceptionBreakpointsFilter[] = null
  supportsSetVariable?: boolean = false;
  supportsRestartFrame?: boolean = false;
  supportsCompletionsRequest?: boolean = false;
  completionTriggerCharacters?: string[] = null
  supportsModulesRequest? = false
  additionalModuleColumns?: ColumnDescriptor[] = null
  supportedChecksumAlgorithms?: ChecksumAlgorithm[] = null
  supportsExceptionOptions? = false
  supportsValueFormattingOptions = false
  supportTerminateDebuggee = false
  supportSuspendDebuggee = false
  supportsDelayedStackTraceLoading = false
  supportsLoadedSourcesRequest = false
  supportsLogPoints? = false
  supportsTerminateThreadsRequest = false
  supportsSetExpression? = false
  supportsTerminateRequest = false
  supportsDataBreakpoints? = false
  supportsReadMemoryRequest? = false
  supportsWriteMemoryRequest? = false
  supportsDisassembleRequest? = false
  supportsCancelRequest? = false
  supportsClipboardContext? = false
  supportsInstructionBreakpoints? = false
  supportsExceptionFilterOptions? = false
  supportsSingleThreadExecutionRequests? = false
}
```

Also, we won't add support for block scoping within Boogie implementations, which would prevent Boogie clients from having to rename variables in some situations to avoid collisions. Renaming variables related to debugging because the changed variable names will show up in the variables window while debugging.

### Implementation
A C# library for implementing the Debug adapter protocol is [here](https://github.com/OmniSharp/csharp-language-server-protocol).

We will update `Microsoft.Boogie.CapturedState` so it includes the source Token of the `AssumeCmd` it originated from, so clients don't need to print this information in the captureState description and parse it again later.

**Comment by robin-aws** [comment]
I agree you shouldn't worry about supporting breakpoints for v1 of this feature, but for the future: could you support user requests for breakpoints just by adding `assume {:captureState} true` statements? Mind you, I'm not yet sure whether arbitrary breakpoints are actually useful in this case, since you're not actually debugging the process of stepping through each statement, you're debugging the state of a static analysis.

**Comment by keyboardDrummer** [comment]
> I agree you shouldn't worry about supporting breakpoints for v1 of this feature, but for the future: could you support user requests for breakpoints just by adding `assume {:captureState} true` statements? Mind you, I'm not yet sure whether arbitrary breakpoints are actually useful in this case, since you're not actually debugging the process of stepping through each statement, you're debugging the state of a static analysis.

Actually, supporting the `setBreakpoints` request will be part of v1. There's no capability for supporting that since it's mandatory. `setFunctionBreakpoints`, which we won't support, allows settings breakpoint based on the name of a function without specifying a source locations. I honestly don't know what it's use-case is, maybe it's for breaking in dependencies.
